File: edd/Clase 5/Set.hs
module Set
    (Set, emptyS, addS, belongs, sizeS, removeS, unionS, setToList)
    where

data Set a = S [a] Int
    deriving Show

emptyS :: Set a -- Crea un conjunto vacío.
emptyS = S [] 0

addS :: Eq a => a -> Set a -> Set a --Dados un elemento y un conjunto, agrega el elemento al conjunto.
addS a set = if belongs a set
                then set
                else agregar a set

agregar :: Eq a => a -> Set a -> Set a
agregar a (S as n) = (S (a:as) (n + 1))

belongs :: Eq a => a -> Set a -> Bool --Dados un elemento y un conjunto indica si el elemento pertenece al conjunto.
belongs a (S xs _) = pertenece a xs 

pertenece :: Eq a => a -> [a] -> Bool
pertenece a []      = False
pertenece a (x:xs)  = (a == x) || pertenece a xs

sizeS :: Eq a => Set a -> Int --Devuelve la cantidad de elementos distintos de un conjunto.
sizeS (S _ n) = n

removeS :: Eq a => a -> Set a -> Set a
removeS a set = if belongs a set
                    then removerDeSet a set
                    else set

removerDeSet :: Eq a => a -> Set a -> Set a --Saca el elemento del set. Debe de existir ese elemento en el set
removerDeSet a (S xs n) = (S (removerDeLista a xs) (n - 1))


removerDeLista :: Eq a => a -> [a] -> [a] --Borra un elemento del conjunto.
removerDeLista a [] = []
removerDeLista a (x:xs) = if (a == x)
                            then xs
                            else x : (removerDeLista a xs)

unionS :: Eq a => Set a -> Set a -> Set a --Dados dos conjuntos devuelve un conjunto con todos los elementos de ambos conjuntos.
unionS (S xs n) set = agregarASet xs set

agregarASet :: Eq a => [a] -> Set a -> Set a
agregarASet [] set      = set
agregarASet (x:xs) set  = agregarASet xs (addS x set)

setToList :: Eq a => Set a -> [a] --Dado un conjunto devuelve una lista con todos los elementos distintos del conjunto.
setToList (S xs n) = xs

File: edd/Clase 11/Set.cpp
#include "Set.h"
#include <iostream>
using namespace std;
#include <stdio.h>

Set emptyS(){
    SetSt* nuevoSet = new SetSt;
    nuevoSet->cantidad = 0;
    nuevoSet->primero = NULL;
    return nuevoSet;
}
// Crea un conjunto vacío.

bool isEmptyS(Set s){
    return s->cantidad == 0;
}
// Indica si el conjunto está vacío.

bool belongsS(int x, Set s){
    NodoS* nodoActual = s->primero;
    while(nodoActual != NULL && nodoActual->elem != x){
        nodoActual = nodoActual->siguiente;
    }
    return nodoActual != NULL;
}
// Indica si el elemento pertenece al conjunto.

void AddS(int x, Set s){
    NodoS* nuevoElemento = new NodoS;
    nuevoElemento->elem = x;
    nuevoElemento->siguiente = NULL;
    if(s->primero == NULL){
        s->primero = nuevoElemento;
        s->cantidad++;
    }else{
        NodoS* nodoActual = s->primero;
        if(nodoActual->elem != x){
            while(nodoActual->siguiente != NULL && nodoActual->siguiente->elem != x){
                nodoActual = nodoActual->siguiente;
            }
            if(nodoActual->siguiente == NULL){
                s->cantidad++;
                nodoActual->siguiente = nuevoElemento;
            }else{
                delete nuevoElemento;
            }
        }
    }
}
// Agrega un elemento al conjunto.

void RemoveS(int x, Set s){
    NodoS* nodoActual = s->primero;
    if(nodoActual != NULL && nodoActual->elem == x){
        delete s->primero;
        s->primero = NULL;
        s->cantidad--;
    }else if(nodoActual != NULL){
        while(nodoActual->siguiente != NULL && nodoActual->siguiente->elem != x){
            nodoActual = nodoActual->siguiente;
            cout << "X";
        }
        if(nodoActual->siguiente != NULL){
            NodoS* nodoABorrar = nodoActual->siguiente;
            nodoActual->siguiente = nodoActual->siguiente->siguiente;
            delete nodoABorrar;
            s->cantidad--;
        }
    }
}
// Quita un elemento dado si lo posee.

int sizeS(Set s){
    return s->cantidad;
}
// Devuelve la cantidad de elementos.
LinkedList setToList(Set s){
    NodoS* nodoActual = s->primero;
    LinkedList nuevaLista = nil();
    while(nodoActual != NULL){
        Snoc(nodoActual->elem, nuevaLista);
        nodoActual = nodoActual->siguiente;
    }
    return nuevaLista;
}
// Devuelve una lista con los lementos del conjunto.

void DestroyS(Set s){
    NodoS* nodoActual = s->primero;
    NodoS* siguienteNodo;
    while(nodoActual != NULL){
        siguienteNodo = nodoActual->siguiente;
        delete nodoActual;
        nodoActual = siguienteNodo;
    }
    delete s;
}
// Libera la memoria ocupada por el conjunto.
